krabs/krabs/ut.hpp

```diff
@@ -99,42 +105,79 @@ namespace krabs { namespace details {
         // for the same GUID are provided and request different provider flags.
         // TODO: Only forward the calls that are requested to each provider.
         for (auto &provider : trace.providers_) {
-            if (providerFlags.find(provider.get().guid_) != providerFlags.end()) {
-                providerFlags[provider.get().guid_] = std::make_tuple<UCHAR, ULONGLONG, ULONGLONG, UCHAR> (0, 0, 0, 0);
+            if (provider_flags.find(provider.get().guid_) != provider_flags.end()) {
+                provider_flags[provider.get().guid_].flags_tuple_ = std::make_tuple<UCHAR, ULONGLONG, ULONGLONG, UCHAR> (0, 0, 0, 0);
             }
 
-            std::get<0>(providerFlags[provider.get().guid_]) |= provider.get().level_;
-            std::get<1>(providerFlags[provider.get().guid_]) |= provider.get().any_;
-            std::get<2>(providerFlags[provider.get().guid_]) |= provider.get().all_;
-            std::get<3>(providerFlags[provider.get().guid_]) |= provider.get().trace_flags_;
+            std::get<0>(provider_flags[provider.get().guid_].flags_tuple_) |= provider.get().level_;
+            std::get<1>(provider_flags[provider.get().guid_].flags_tuple_) |= provider.get().any_;
+            std::get<2>(provider_flags[provider.get().guid_].flags_tuple_) |= provider.get().all_;
+            std::get<3>(provider_flags[provider.get().guid_].flags_tuple_) |= provider.get().trace_flags_;
+
+            for(const auto& filter : provider.get().filters_)
+            {
+                if (filter.provider_filter_event_id() > 0)
+                {
+                	//native id existing, set native filters
+                	provider_flags[provider.get().guid_].provider_filter_event_ids_.push_back(filter.provider_filter_event_id());
+                }
+            }
         }
 
-        for (auto &provider : providerFlags) {
-            GUID guid = provider.first;
-
+        for (auto &provider : provider_flags) {
+            //compose native event params by native events ids 
             ENABLE_TRACE_PARAMETERS parameters;
+            ZeroMemory(&parameters, sizeof(parameters));
             parameters.Version = ENABLE_TRACE_PARAMETERS_VERSION_2;
             parameters.SourceId = provider.first;
-            parameters.EnableProperty = std::get<3>(provider.second);
-            parameters.ControlFlags = 0;
+            
+            GUID guid = provider.first;
+            parameters.EnableProperty = std::get<3>(provider.second.flags_tuple_);
             parameters.EnableFilterDesc = nullptr;
             parameters.FilterDescCount = 0;
+            EVENT_FILTER_DESCRIPTOR filterDesc;
+            std::unique_ptr<BYTE[]> filterMemoryPtr;
+
+            if (provider.second.provider_filter_event_ids_.size() > 0)
+            {
+                //event filters existing, se native filters using API
+                parameters.FilterDescCount = 1;  
+
+                ZeroMemory(&filterDesc, sizeof(filterDesc));
```

Comment:
consider pulling this block out into a separate method that sets up `filterDesc` appropriately

Reply:
I would not do it for simplicity e.g. because of scoping of  allocated std::unique_ptr<BYTE[]> filterMemoryPtr memory which will be allocated inside this method but released outside... 
Please let me know if you insist on this change request and i will do it...
Thanks